clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -146,8 +152,9 @@ private void onFailure(final Throwable exception, final long responseTimeMs) {
         logger.debug("failed sending heartbeat due to {}", exception.getMessage());
     }
 
-    private void onSuccess(final ConsumerGroupHeartbeatResponse response, long currentTimeMs) {
+    private void onResponse(final ConsumerGroupHeartbeatResponse response, long currentTimeMs) {
```

Comment:
Again brainstorming based on the [draft](https://github.com/apache/kafka/pull/14413), this would be much simplified with the move of the error handling more into the MembershipManager I expect. Here, instead of having to paths to update state (now there are 2 calls, one to membershipManager.updateState and another for all the error handling), we could simply have something like:

```
private void onResponse(final ConsumerGroupHeartbeatResponse response, long currentTimeMs) {
        if (response.data().errorCode() == Errors.NONE.code()) {
            // Heartbeat manager specifics for success - nothing affecting state
        } else {
            // Heartbeat manager specifics for failure - nothing affecting state & no error handling other than identifying when coordinatorRequestManager.markCoordinatorUnknown is needed, which should be here I think
            }
        }
        // Update state - single point of interaction with the membershipMgr
        Optional<Errors> error = membershipManager.updateState(response.data());
        if (error.isPresent()) {
            nonRetriableErrorHandler.handle(error.get().exception());
        }
    }
```

Reply:
Why don't we let the heartbeat request manager to handle all the errors?  Technically, these are heartbeat errors, and it would be more centralized to handle them on the manager.

The membershipManager really could just ensure if the transition is valid.

Reply:
Sounds good, actually better to move it all to the Heartbeat manager, given that it is the more concerned about the HB errors. The membershipMgr in the end only needs to know about what affects the state (success, fencing and fatal failures)